storage/aptosdb/src/db_debugger/truncation/mod.rs

```diff
@@ -0,0 +1,342 @@
+use aptos_jellyfish_merkle::{node_type::NodeKey, StaleNodeIndex};
+use aptos_schemadb::{
+    schema::{Schema, SeekKeyCodec},
+    ReadOptions, SchemaBatch, DB,
+};
+use aptos_types::transaction::Version;
+use clap::Parser;
+use std::{path::PathBuf, sync::Arc};
+
+#[derive(Parser)]
+#[clap(about = "Delete all data after the provided version.")]
+pub struct Cmd {
+    #[clap(long, parse(from_os_str))]
+    db_dir: PathBuf,
+
+    target_version: u64,
+
+    ledger_db_batch_size: usize,
+}
+
+impl Cmd {
+    pub fn run(self) -> Result<()> {
+        let (ledger_db, state_merkle_db, _kv_db) = AptosDB::open_dbs(
+            &self.db_dir,
+            RocksdbConfigs::default(),
+            /*readonly=*/ false,
+        )?;
+
+        // TODO(grao): Handle kv db once we enable it.
+
+        let ledger_db_version = Self::get_current_version_in_ledger_db(&ledger_db)?
+            .expect("Current version of ledger db must exist.");
+        let state_merkle_db_version =
+            Self::get_current_version_in_state_merkle_db(&state_merkle_db)?
+                .expect("Current version of state merkle db must exist.");
+
+        assert!(self.target_version <= ledger_db_version);
+        assert!(ledger_db_version >= state_merkle_db_version);
+
+        let state_merkle_target_version =
+            Self::find_tree_root_at_or_before(&ledger_db, &state_merkle_db, self.target_version)?
+                .expect(&format!(
+                    "Could not find a valid root before or at version {}, maybe it was pruned?",
+                    self.target_version
+                ));
+
+        Self::truncate_state_merkle_db(&state_merkle_db, state_merkle_target_version)?;
+
+        let ledger_db = Arc::new(ledger_db);
+        Self::truncate_ledger_db(
+            Arc::clone(&ledger_db),
+            ledger_db_version,
+            self.target_version,
+            self.ledger_db_batch_size,
+        )?;
+
+        if state_merkle_target_version < self.target_version {
+            StateStore::catch_up_state_merkle_db(Arc::clone(&ledger_db), state_merkle_db)?;
+        }
+
+        Ok(())
+    }
+
+    fn get_current_version_in_ledger_db(ledger_db: &DB) -> Result<Option<Version>> {
+        let mut iter = ledger_db.iter::<TransactionInfoSchema>(ReadOptions::default())?;
+        iter.seek_to_last();
+        Ok(iter.next().transpose()?.map(|item| item.0))
+    }
+
+    fn get_current_version_in_state_merkle_db(state_merkle_db: &DB) -> Result<Option<Version>> {
+        Self::find_closest_node_version_at_or_before(state_merkle_db, u64::max_value())
+    }
+
+    fn find_tree_root_at_or_before(
+        ledger_db: &DB,
+        state_merkle_db: &DB,
+        version: Version,
+    ) -> Result<Option<Version>> {
+        match Self::find_closest_node_version_at_or_before(state_merkle_db, version)? {
+            Some(closest_version) => {
+                if Self::root_exist_at_version(state_merkle_db, closest_version)? {
```

Comment:
what is the reason that the root might not exist? due to pruning?

Reply:
Yes. There might be a case that a root node is pruned, but the non-root node with the same version is not because it is referenced by a later version root node.